### ISSUE TITLE
李航机器学习之集成学习

### ISSUE THREAD

**Issue by MiaoRain** [MiaoRain/lund#21]
**https://zhuanlan.zhihu.com/p/27126737** **adaboost详解**
http://sofasofa.io/forum_main_post.php?postid=1003214  **adaboost做回归预测**
https://zhuanlan.zhihu.com/p/39972832
![image](https://user-images.githubusercontent.com/47746159/81715469-8d8c7400-9478-11ea-8429-45fe9e0ae61c.png)
理论上任何学习器都可以用于Adaboost.但一般来说，使用最广泛的Adaboost弱学习器是决策树和神经网络。对于决策树，Adaboost分类用了CART分类树，而Adaboost回归用了CART回归树。

Adaboost的主要优点有：
1. Adaboost作为分类器时，分类精度很高
2. 在Adaboost的框架下，可以使用各种回归分类模型来构建弱学习器，非常灵活。
3. 作为简单的二元分类器时，构造简单，结果可理解。
4. 不容易发生过拟合

Adaboost的主要缺点有：对异常样本敏感，异常样本在迭代中可能会获得较高的权重，影响最终的强学习器的预测准确性。

![image](https://user-images.githubusercontent.com/47746159/79050819-d5776b80-7c2c-11ea-8b3a-fc464828dcda.png)
![image](https://user-images.githubusercontent.com/47746159/79050924-867e0600-7c2d-11ea-9e96-e59880dbf419.png)
训练误差以指数减少
![image](https://user-images.githubusercontent.com/47746159/79051351-01482080-7c30-11ea-811f-cc99df9ece8d.png)
![image](https://user-images.githubusercontent.com/47746159/79051393-53894180-7c30-11ea-8e45-df3ebad9a777.png)

**Comment by MiaoRain** [comment]
**提升树**
![image](https://user-images.githubusercontent.com/47746159/79051631-6b14fa00-7c31-11ea-99f7-26f6ba9f01b3.png)
![image](https://user-images.githubusercontent.com/47746159/79051766-54bb6e00-7c32-11ea-84f1-80e3cff3db0c.png)
![image](https://user-images.githubusercontent.com/47746159/79051958-7a954280-7c33-11ea-9104-f063d8dee96b.png)

**Comment by MiaoRain** [comment]
https://zhuanlan.zhihu.com/p/35796662

**Comment by MiaoRain** [comment]
Week5【任务1】第八章前向分步算法
![image](https://user-images.githubusercontent.com/47746159/79052748-e5954800-7c38-11ea-91a0-cbc6fec1a974.png)

**Comment by MiaoRain** [comment]
week5【任务3】第八章作业讲解-提升方法
![image](https://user-images.githubusercontent.com/47746159/79052772-11183280-7c39-11ea-91f6-6b7c2bf37154.png)
![image](https://user-images.githubusercontent.com/47746159/79052832-753af680-7c39-11ea-91b1-8b60741e65ab.png)

**Comment by MiaoRain** [comment]
#```
 自编程求解例8.1

import numpy as np

class AdaBoost:
    def __init__(self,X,y,tol=0.05,max_iter=10):
        self.X=X   # 训练数据 实例
        self.y=y   # 训练数据 标签
        self.tol=tol # 训练中止条件 right_rate>self.tol
        self.max_iter=max_iter # 最大迭代次数
        self.w=np.full((X.shape[0]),1/X.shape[0]) #初始化样本权重w
        self.alpha=[]  # 弱分类器权重
        self.G=[] # 弱分类器
        self.min_v=min(X)-0.5 #分类阈值下届
        self.max_v=max(X)+0.5 #分类阈值上届

    def _class(self):
        """以带权重的分类误差最小为目标，选择最佳分类阈值"""
        e_min=np.inf   # e_min 分类误差
        v_best=None    # v_best 最佳分类阈值
        sign=None      # sign 小于分类阈值的样本属于的标签类别
        for v in np.arange(self.min_v,self.max_v+0.5,1):
            # 遍历可能v_best可取值，寻找最优解
            e_1=-(self.y[self.X<v]-1)*self.w[self.X<v] # 假设小于阈值 分类为1 获取分类误差*2
            e_2=(self.y[self.X>v]+1)*self.w[self.X>v]  # 假设大于阈值 分类为-1 获取分类误差*2
            e=(e_1.sum()+e_2.sum())/2  # 计算整个分类误差
            if e<0.5:   # 若分类误差小于0.5 说明X<v y->1
                flag=1
            else:       # 若分类误差大于0.5 取其反向 说明X<v y->-1
                e=1-e
                flag=-1
            if e<e_min: # 保留最优解
                e_min=e
                sign=flag
                v_best=v
        return v_best,sign,e_min

    def updata_w(self):
        """更新样本权重w"""
        v,sign=self.G[-1]  # 以下2行 根据上一轮的弱分类器更新样本权重
        alpha=self.alpha[-1]
        G=np.zeros(self.y.size,dtype=int) # 以下三行重建弱分类器
        G[self.X<v]=sign
        G[self.X>v]=-sign
        # G_1=np.full((np.where(self.X<v))[0].shape[0],sign)
        # G_2=np.full((np.where(self.X>v))[0].shape[0],-sign)
        # G=np.hstack([G_1,G_2])
        P=self.w*np.exp(-alpha*self.y*G) #以下2行根据公式8.4 8.5 更新样本权重
        self.w=P/P.sum()

    def base_estimator(self,X,i):
        """计算单个弱分类器预测输出"""
        v,sign = self.G[i]
        _G_1 = np.full((np.where(X<v))[0].shape[0], sign)
        _G_2 = np.full((np.where(X>v))[0].shape[0], -sign)
        _G = np.hstack([_G_1, _G_2])
        return _G

    def fit(self):
        """对训练数据进行学习"""
        G=0
        for i in range(self.max_iter):
            class_v,sign,e=self._class() # 获取当前迭代最佳分类阈值
            alpha=1/2*np.log((1-e)/e)   # 计算本轮弱分类器的系数
            self.alpha.append(alpha)    # 保存弱分类器系数
            self.G.append((class_v,sign)) # 保存弱分类器
            _G=self.base_estimator(self.X,i) # 以下4行计算当前总分类器（之前所有弱分类器加权和）分类效率
            G+=alpha*_G
            y_predict=np.sign(G)
            error_rate=np.sum(np.abs(y_predict-self.y))/2/self.y.shape[0]
            if error_rate<self.tol: # 满足中止条件 则跳出循环
                print("迭代次数:",i+1)
                break
            else:
                self.updata_w()   # 若不满足，更新权重，继续迭代

    def predict(self,X):
        """对新数据进行预测"""
        G=0
        for i in range(len(self.alpha)):
            # 遍历每一个弱分类器，进行加权
            _G=self.base_estimator(X,i)
            alpha = self.alpha[i]
            G+=alpha*_G
        y_predict=np.sign(G)
        return y_predict.astype(int)

    def score(self,X,y):
        """对训练效果进行评价"""
        y_predict=self.predict(X)
        error_rate=np.sum(np.abs(y_predict-y))/2/y.shape[0]
        return 1-error_rate

def main():
    X=np.array([0,1,2,3,4,5,6,7,8,9])
    y=np.array([1,1,1,-1,-1,-1,1,1,1,-1])
    clf=AdaBoost(X,y)
    clf.fit()
    y_predict=clf.predict(X)
    score=clf.score(X,y)
    print("原始输出:",y)
    print("预测输出:",y_predict)
    print("预测正确率：{:.2%}".format(score))

if __name__=="__main__":
    main()
```

**Comment by MiaoRain** [comment]
```
# 自编程求解例8.1
from sklearn.ensemble import AdaBoostClassifier
import numpy as np

def main():
    # X=np.array([0,1,2,3,4,5,6,7,8,9]).reshape(10,1)
    # y=np.array([1,1,1,-1,-1,-1,1,1,1,-1])
    X=np.array([[0,1,3],
                [0,3,1],
                [1,2,2],
                [1,1,3],
                [1,2,3],
                [0,1,2],
                [1,1,2],
                [1,1,1],
                [1,3,1],
                [0,2,1]
               ])
    y=np.array([-1,-1,-1,-1,-1,-1,1,1,-1,-1])
    clf=AdaBoostClassifier()
    clf.fit(X,y)
    y_predict=clf.predict(X)
    score=clf.score(X,y)
    print("原始输出:",y)
    print("预测输出:",y_predict)
    print("预测正确率：{:.2%}".format(score))

if __name__=="__main__":
    main()
```

**Comment by MiaoRain** [comment]
**七、对缺失值的处理**

xgboost把缺失值当做稀疏矩阵来对待，本身的在节点分裂时不考虑的缺失值的数值。缺失值数据会被分到左子树和右子树分别计层损失，选择较优的那一个。如果训练中没有数据缺失，预测时出现了数据缺失，那么默认被分类到右子树。
八、xgboost的改进

1、目标函数通过二阶泰勒展开式做近似 ；

2、定义了树的复杂度，并应用到目标函数中 ；

3、分裂结点处通过结构打分和分割损失动态生长 ；

4、可以处理稀疏、缺失数据 ；

5、可以通过特征的列采样防止过拟合；

**九、xgboost参数**

XGBoost的参数可以分为三种类型：通用参数、booster参数以及学习目标参数。

General parameters：参数控制在提升（boosting）过程中使用哪种booster，常用的booster有树模型（tree）和线性模型（linear model）。
Booster parameters：这取决于使用哪种booster。
Learning Task parameters：控制学习的场景，例如在回归问题中会使用不同的参数控制排序。
除了以上参数还可能有其它参数，在命令行中使用。
xgboost中常用参数的意义:

booster： gbtree使用基于树的模型进行提升计算,gblinear使用线性模型进行提升计算。

silent: 取0时表示打印出运行时信息,取1时表示以缄默方式运行。

objective：定义学习任务及相应的学习目标，可选的目标函数如下：

“reg:linear” –线性回归。
“reg:logistic” –逻辑回归。
“binary:logistic” –二分类的逻辑回归问题，输出为概率。
“binary:logitraw” –二分类的逻辑回归问题，输出的结果为wTx。
“count:poisson” –计数问题的poisson回归，输出结果为poisson分布。 在poisson回归中，max_delta_step的缺省值为0.7。(used to safeguard optimization)
“multi:softmax” –让XGBoost采用softmax目标函数处理多分类问题，同时需要设置参数num_class（类别个数）
“multi:softprob” –和softmax一样，但是输出的是ndata * nclass的向量，可以将该向量reshape成ndata行nclass列的矩阵。没行数据表示样本所属于每个类别的概率。
“rank”:pairwise –set XGBoost to do ranking task by minimizing the pairwise loss
eval_metric: 校验数据所用评价指标。

gamma: 在树的叶子节点上进一步划分所需的最小损失减少。数值越大,算法越保守。默认0。

min_child_weight: 在线性回归模型中,这个参数是指建立每个模型所需要的最小样本数。调大这个参数能够控制过拟合,默认1。

max_depth:树的最大深度，默认为6。

lambda:L2正则的惩罚系数。

subsample:用于训练模型的子样本占整个样本集合的比例。防止过拟合。

colsample_bytree:1,在建立树时对特征采样的比例。

eta: 通常最后设置eta为0.01~0.2，eta通过缩减特征的权重使提升计算过程更加保守。

nthread:XGBoost运行时的线程数。默认值是当前系统可以获得的最大线程数。

**Comment by MiaoRain** [comment]
https://github.com/dmlc/xgboost  **xgboost**
https://xgboost.readthedocs.io/en/latest/python/index.html **xgboost官网**
https://zhuanlan.zhihu.com/p/91817667 **xgboost 知乎**
https://blog.csdn.net/tyhj_sf/article/details/84954120  **xgboost 原理详解**
https://www.bilibili.com/video/BV1Ks411V7Jg?p=4 **xgboost B站视频**
https://www.jianshu.com/p/7467e616f227 **xgboost的原理没你想像的那么难**
https://zhuanlan.zhihu.com/p/81368182 **[校招-基础算法]GBDT/XGBoost常见问题** 
![image](https://user-images.githubusercontent.com/47746159/81733711-bfaacf80-9492-11ea-8200-28856e4f6ac3.png)

https://lightgbm.readthedocs.io/en/latest/index.html  **lightBGM**
https://blog.csdn.net/weixin_39807102/article/details/81912566 **LightGBM算法总结**
https://blog.csdn.net/shine19930820/article/details/79123216 **『 论文阅读』LightGBM原理**
LightGBM
简单介绍
LightGBM是一个实现GBDT算法的分布式高效框架。它通过leaf-wise分裂方法进行决策树的生成，通过基于直方图的算法寻找特征分割点，并支持并行学习，能够更高效的处理大数据，也得到了越来越广泛的应用。

更深一步的介绍
首先，它通过**leaf-wise分裂**方法产生比level-wise分裂方法更复杂的树，能够实现更高的准确率。虽然这样有时候会导致过拟合，但可以通过设置 max-depth 参数来防止过拟合的发生。（每一次的生长都是选取分裂增益最高的节点，而不是对一层中的所有节点都进行分裂）。
其次，它使用基于**直方图的算法**，将连续的特征值分桶(buckets)装进离散的箱子(bins)，并通过直方图做差加速计算兄弟节点的直方图，能够加速训练过程，并实现更少的内存占用。
另外，它支持**并行学习，包括特征并行和数据并行**
特征并行的主要思想是不同机器在不同的特征集合上分别寻找最优的分割点，然后在机器间同步最优的分割点。（mapreduce思想）
数据并行则是让不同的机器先在不同的记录集合上构造直方图，然后进行全局的合并，最后在合并的直方图上面寻找最优分割点。
它还支持直接输入类别特征，在对离散类别特征分裂时，每个取值都当作一个桶，分裂时的增益算的是”是否属于某个类别“的增益，对于类别特征的**操作类似于one-hot编码**。
![image](https://user-images.githubusercontent.com/47746159/81741381-c8090780-949e-11ea-9ebf-3b672d3447ef.png)


https://github.com/catboost/catboost  **catboost** 
https://blog.csdn.net/program_developer/article/details/104263623 **深入理解CatBoost**

**Comment by MiaoRain** [comment]
https://zhuanlan.zhihu.com/p/61678619
**xgb是对gbdt的优化改进**



1、目标函数中加入了**正则项**来控制模型的复杂度，替代原来的剪枝方法。



2、利用了**one-hot编码**等情况中的特征稀疏性。（**仅对特征值为非缺失值的样本的特征进行遍历**）



3、支持列抽样（同random forest）。



4、数据事先排序并按block结构保存，有利于并行运算（树的生成还是串行的，这里说的**并行计算指并行计算各个特征的增益或是基尼系数**）。除此之外，xgb还通过一种**可并行的近似直方图算法来高效生成候选的分割点**。



5、对损失函数进行了优化，gbdt只用到了其一阶导数信息，而x**gb同时用到其一阶导与二阶导**。



**lgb则是在xgb基础上进一步改进**



1、内存需求小：x**gb使用基于pre-sorted的决策树算**法，而**lgb使用基于histogram的决策树算法**。histogram算法占用的内存很小：pre-sorted需要两倍数据大小的内存空间，一半用于数据（float32），一半用于存放排好序的索引，而histogram不需要存放索引，且特征值只需要存放离散后的值，用uint8即可，故内存需求仅为**pre-sorted的1/8**。



2、计算速度快：决策树算法的主要操作包括“**寻找分割点**”与“**数据分割**”两步，pre-sorted算法和histogram算法在“寻找分割点”上的时间复杂度是一致的；但是在“数据分割”上histogram要快，histogram所有特征共享一张索引，而pre-sorted一个特征对应一张索引，若集合level-wise，pre-sorted也可以共用一张索引，但是会带来很多随机访问的问题，速度仍不及histogram。此外，histogram算法还减少了计算分割点增益的次数。



3、通信代价小：**histogram算法的通信代价远远小于pre-sorted**，可用于分布式计算。



但是，histogram**不能找到很精确的分割点**，训练误差不如pre-sorted算法，可以说是牺牲一定精度来换取速度。需要指出的是，这种**粗犷的分割相当于自带正则效果**，所以测试集的误差两种决策树算法差距不大。

**关键两点差别**



**1、决策树算法**



XGBoost使用的是**pre-sorted**算法，能够更精确的找到数据分隔点；LightGBM使用的是**histogram算法**，占用的内存更低，数据分隔的复杂度更低。



2、**决策树生长策略**



XGBoost采用的是level（depth）-wise生长策略，**能够同时分裂同一层的叶子**，从而进行多线程优化，不容易过拟合；但不加区分的对待同一层的叶子，带来了很多没必要的开销。



LightGBM采用leaf-wise生长策略，每次从**当前所有叶子中找到分裂增益最大**（一般也是数据量最大）的一个叶子，然后分裂，如此循环；但会生长出比较深的决策树，产生过拟合。因此，LightGBM 在leaf-wise 之上增加了一个最大深度的限制，在保证高效率的同时防止过拟合

**Comment by MiaoRain** [comment]
**集成学习之模型融合与随机森林(Random Forest)**
https://blog.csdn.net/Dby_freedom/article/details/82057303
